File: ts/src/resources/chat-ai.ts
import { HttpTransport } from "../http.js"

// ─── AI Agents ────────────────────────────────────────────────────────────────

export interface AiAgent {
  id: string
  name: string
  description?: string
  model?: string
  instructions?: string
  [key: string]: unknown
}

export interface CreateAiAgentInput {
  name: string
  workflow_name: string
  /**
   * The provider this AI agent chats through. Use `"system"` to delegate to
   * the org's default LLM provider; pass a specific provider UUID to pin one.
   */
  provider_id: string
  /**
   * The model name. With `provider_id: "system"` use a model name (e.g.
   * `"gpt-4o"`) or the literal `"Default"` to fall back to the system default.
   * For a custom provider, pass that provider's model id.
   */
  model_id: string
  description?: string
  /** @deprecated legacy field — chat orchestrator type, leave unset. */
  model?: string
  instructions?: string
  [key: string]: unknown
}

// ─── Document AI ──────────────────────────────────────────────────────────────

export interface DocumentAIInput {
  modelName: string
  url: string
  organizationId: string
  boardId?: string
  language?: string
  additionalInstructions?: string
  additionalDocumentInstructions?: string
  processModelName?: string
  fileUrlToFill?: string
  tools?: Record<string, unknown>[]
  utc?: number
  chunkSize?: number
  maxConcurrent?: number
  maxRetries?: number
  useEnhancedProcessing?: boolean
}

export interface DocumentAIResponse {
  success: boolean
  data: Record<string, unknown> & { filledPdfUrl?: string }
}

/**
 * Imbrace-specific endpoints exposed by the chat-ai service. The
 * upstream OpenWebUI surface (chats, files list/upload, audio, knowledge,
 * folders, prompts, tools, models) is intentionally not wrapped here:
 * those endpoints require an OpenWebUI session JWT (issued only by the
 * OpenWebUI login flow) and reject `x-api-key` / `x-access-token` auth.
 *
 * What this resource covers:
 *   • Agent file upload and content extraction
 *   • Document AI (vision-model document processing) and provider listing
 *   • Imbrace AI agents (`/accounts/assistants`, `/assistant_apps`, …)
 */
export class ChatAiResource {
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  // ─── Files (Imbrace additions on top of OpenWebUI) ──────────────────────────

  /** Upload an agent-specific file. Endpoint: /ai/v3/files/agent */
  async uploadAgentFile(body: FormData): Promise<unknown> {
    return this.http.getFetch()(`${this.base}/files/agent`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  /** Extract content from an uploaded file (PDF/etc). Endpoint: /ai/v3/files/extract */
  async extractFile(body: FormData): Promise<unknown> {
    return this.http.getFetch()(`${this.base}/files/extract`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  // ─── Document AI ──────────────────────────────────────────────────────────

  /** Process a document with a vision model and extract structured data. Endpoint: /ai/v3/document/ */
  async processDocument(body: DocumentAIInput): Promise<DocumentAIResponse> {
    return this.http.getFetch()(`${this.base}/document/`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  /** List LLM providers configured for the org — these are the models available for document AI. Endpoint: /ai/v3/providers */
  async listDocumentModels(): Promise<unknown[]> {
    return this.http.getFetch()(`${this.base}/providers`, { method: "GET" }).then(r => r.json())
  }

  // ─── AI Agents ─────────────────────────────────────────────────────────────

  async listAiAgents(): Promise<AiAgent[]> {
    return this.http.getFetch()(`${this.base}/accounts/assistants`, { method: "GET" }).then(r => r.json())
  }

  async getAiAgent(id: string): Promise<AiAgent> {
    return this.http.getFetch()(`${this.base}/assistants/${id}`, { method: "GET" }).then(r => r.json())
  }

  async createAiAgent(body: CreateAiAgentInput): Promise<AiAgent> {
    if (!body?.provider_id || !body?.model_id) {
      throw new Error(
        "createAiAgent: provider_id and model_id are required. " +
        "Use { provider_id: \"system\", model_id: \"gpt-4o\" } for the system default, " +
        "or pass a custom provider's UUID and model name.",
      )
    }
    return this.http.getFetch()(`${this.base}/assistant_apps`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateAiAgent(id: string, body: Partial<CreateAiAgentInput>): Promise<AiAgent> {
    return this.http.getFetch()(`${this.base}/assistant_apps/${id}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteAiAgent(id: string): Promise<boolean> {
    const r = await this.http.getFetch()(`${this.base}/assistant_apps/${id}`, { method: "DELETE" })
    return r.ok
  }

  async updateAiAgentInstructions(id: string, instructions: string): Promise<AiAgent> {
    return this.http.getFetch()(`${this.base}/assistants/${id}/instructions`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ instructions }),
    }).then(r => r.json())
  }

  async listAiAgentSubAgents(): Promise<unknown[]> {
    return this.http.getFetch()(`${this.base}/assistants/agents`, { method: "GET" }).then(r => r.json())
  }
}

File: py/src/imbrace/resources/chat_ai.py
from typing import Any, Dict, List, Optional
from ..http import HttpTransport, AsyncHttpTransport


# Imbrace-specific endpoints exposed by the chat-ai service. The upstream
# OpenWebUI surface (chats, files list/upload, audio, knowledge, folders,
# prompts, tools, models) is intentionally not wrapped here: those endpoints
# require an OpenWebUI session JWT (issued only by the OpenWebUI login flow)
# and reject ``x-api-key`` / ``x-access-token`` auth.


class ChatAiResource:
    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    # --- Files (Imbrace additions on top of OpenWebUI) ---

    def upload_agent_file(self, files: Any, agent_id: str) -> Dict[str, Any]:
        """Upload an agent-specific file."""
        return self._http.request(
            "POST", f"{self._base}/files/agent", files=files, data={"agent_id": agent_id}
        ).json()

    def extract_file(self, files: Any) -> Dict[str, Any]:
        """Extract content from an uploaded file (PDF/etc)."""
        return self._http.request("POST", f"{self._base}/files/extract", files=files).json()

    # --- Document AI ---

    def process_document(
        self,
        model_name: str,
        url: str,
        organization_id: str,
        *,
        board_id: Optional[str] = None,
        language: Optional[str] = None,
        additional_instructions: Optional[str] = None,
        additional_document_instructions: Optional[str] = None,
        process_model_name: Optional[str] = None,
        file_url_to_fill: Optional[str] = None,
        tools: Optional[List[Dict[str, Any]]] = None,
        utc: Optional[int] = None,
        chunk_size: Optional[int] = None,
        max_concurrent: Optional[int] = None,
        max_retries: Optional[int] = None,
        use_enhanced_processing: Optional[bool] = None,
    ) -> Dict[str, Any]:
        """Process a document with a vision model and extract structured data."""
        body: Dict[str, Any] = {"modelName": model_name, "url": url, "organizationId": organization_id}
        if board_id is not None: body["boardId"] = board_id
        if language is not None: body["language"] = language
        if additional_instructions is not None: body["additionalInstructions"] = additional_instructions
        if additional_document_instructions is not None: body["additionalDocumentInstructions"] = additional_document_instructions
        if process_model_name is not None: body["processModelName"] = process_model_name
        if file_url_to_fill is not None: body["fileUrlToFill"] = file_url_to_fill
        if tools is not None: body["tools"] = tools
        if utc is not None: body["utc"] = utc
        if chunk_size is not None: body["chunkSize"] = chunk_size
        if max_concurrent is not None: body["maxConcurrent"] = max_concurrent
        if max_retries is not None: body["maxRetries"] = max_retries
        if use_enhanced_processing is not None: body["useEnhancedProcessing"] = use_enhanced_processing
        return self._http.request("POST", f"{self._base}/document/", json=body).json()

    def list_document_models(self) -> List[Dict[str, Any]]:
        """List LLM providers configured for the org — models available for document AI."""
        return self._http.request("GET", f"{self._base}/providers").json()

    # --- Assistants ---

    def list_assistants(self) -> List[Dict[str, Any]]:
        """List all assistants for the account."""
        return self._http.request("GET", f"{self._base}/accounts/assistants").json()

    def get_assistant(self, assistant_id: str) -> Dict[str, Any]:
        """Get assistant by UUID."""
        return self._http.request("GET", f"{self._base}/assistants/{assistant_id}").json()

    def create_assistant(self, body: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new assistant. Required: name, workflow_name, provider_id, model_id."""
        return self._http.request("POST", f"{self._base}/assistant_apps", json=body).json()

    def update_assistant(self, assistant_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        """Update assistant by UUID."""
        return self._http.request("PUT", f"{self._base}/assistant_apps/{assistant_id}", json=body).json()

    def delete_assistant(self, assistant_id: str) -> bool:
        """Delete assistant by UUID."""
        r = self._http.request("DELETE", f"{self._base}/assistant_apps/{assistant_id}")
        return r.is_success

    def update_assistant_instructions(self, assistant_id: str, instructions: str) -> Dict[str, Any]:
        """Update only the instructions field of an assistant."""
        return self._http.request(
            "PATCH",
            f"{self._base}/assistants/{assistant_id}/instructions",
            json={"instructions": instructions},
        ).json()

    def list_assistant_agents(self) -> List[Dict[str, Any]]:
        """List all assistant agents."""
        return self._http.request("GET", f"{self._base}/assistants/agents").json()


class AsyncChatAiResource:
    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    # --- Files (Imbrace additions on top of OpenWebUI) ---

    async def upload_agent_file(self, files: Any, agent_id: str) -> Dict[str, Any]:
        res = await self._http.request(
            "POST", f"{self._base}/files/agent", files=files, data={"agent_id": agent_id}
        )
        return res.json()

    async def extract_file(self, files: Any) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._base}/files/extract", files=files)
        return res.json()

    # --- Document AI ---

    async def process_document(
        self,
        model_name: str,
        url: str,
        organization_id: str,
        *,
        board_id: Optional[str] = None,
        language: Optional[str] = None,
        additional_instructions: Optional[str] = None,
        additional_document_instructions: Optional[str] = None,
        process_model_name: Optional[str] = None,
        file_url_to_fill: Optional[str] = None,
        tools: Optional[List[Dict[str, Any]]] = None,
        utc: Optional[int] = None,
        chunk_size: Optional[int] = None,
        max_concurrent: Optional[int] = None,
        max_retries: Optional[int] = None,
        use_enhanced_processing: Optional[bool] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"modelName": model_name, "url": url, "organizationId": organization_id}
        if board_id is not None: body["boardId"] = board_id
        if language is not None: body["language"] = language
        if additional_instructions is not None: body["additionalInstructions"] = additional_instructions
        if additional_document_instructions is not None: body["additionalDocumentInstructions"] = additional_document_instructions
        if process_model_name is not None: body["processModelName"] = process_model_name
        if file_url_to_fill is not None: body["fileUrlToFill"] = file_url_to_fill
        if tools is not None: body["tools"] = tools
        if utc is not None: body["utc"] = utc
        if chunk_size is not None: body["chunkSize"] = chunk_size
        if max_concurrent is not None: body["maxConcurrent"] = max_concurrent
        if max_retries is not None: body["maxRetries"] = max_retries
        if use_enhanced_processing is not None: body["useEnhancedProcessing"] = use_enhanced_processing
        res = await self._http.request("POST", f"{self._base}/document/", json=body)
        return res.json()

    async def list_document_models(self) -> List[Dict[str, Any]]:
        res = await self._http.request("GET", f"{self._base}/providers")
        return res.json()

    # --- Assistants ---

    async def list_assistants(self) -> List[Dict[str, Any]]:
        res = await self._http.request("GET", f"{self._base}/accounts/assistants")
        return res.json()

    async def get_assistant(self, assistant_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._base}/assistants/{assistant_id}")
        return res.json()

    async def create_assistant(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._base}/assistant_apps", json=body)
        return res.json()

    async def update_assistant(self, assistant_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PUT", f"{self._base}/assistant_apps/{assistant_id}", json=body)
        return res.json()

    async def delete_assistant(self, assistant_id: str) -> bool:
        r = await self._http.request("DELETE", f"{self._base}/assistant_apps/{assistant_id}")
        return r.is_success

    async def update_assistant_instructions(self, assistant_id: str, instructions: str) -> Dict[str, Any]:
        res = await self._http.request(
            "PATCH",
            f"{self._base}/assistants/{assistant_id}/instructions",
            json={"instructions": instructions},
        )
        return res.json()

    async def list_assistant_agents(self) -> List[Dict[str, Any]]:
        res = await self._http.request("GET", f"{self._base}/assistants/agents")
        return res.json()
